core/src/main/scala/org/apache/spark/rpc/netty/NettyRpcEnv.scala

```diff
@@ -120,9 +120,15 @@ private[netty] class NettyRpcEnv(
       RpcEndpointVerifier.NAME, new RpcEndpointVerifier(this, dispatcher))
   }
 
-  @Nullable
   override lazy val address: RpcAddress = {
-    if (server != null) RpcAddress(host, server.getPort()) else null
+    // When running in client mode (i.e.: not opening a listening port, but connecting to a
+    // 'server' with an open port), the server value is null.
+    // But a valid host still has to be provided, as it is needed later on.
+    // Previously, the server side of the connection was attempting to guess the client
+    // hostname, using the connection information. But the value generated is wrong when
+    // the connection is NATed.
+    // [SPARK-14849]
+    if (server != null) RpcAddress(host, server.getPort()) else RpcAddress(host, -1)
```

Comment:
`-1` is actually confusing. All of executors in the same node will have the same address.